File: LatamDigest/Services/OpenAIService.swift
import Foundation

/// A helper service that wraps calls to the OpenAI API to summarise and
/// cluster news headlines.  This class uses the `Responses` endpoint
/// introduced in 2025 which supports web search, structured outputs and
/// JSON schemas.  You should not call this service directly from the
/// iOS client in production; instead, your backend should perform these
/// calls on a schedule and cache the results.
final class OpenAIService {
    static let shared = OpenAIService()
    private init() {}

    enum ConfigurationError: LocalizedError {
        case missingAPIKey

        var errorDescription: String? {
            switch self {
            case .missingAPIKey:
                return "OPENAI_API_KEY is missing. Configure it in the app environment or move this call to your backend."
            }
        }
    }

    private let baseURL = URL(string: "https://api.openai.com/v1/responses")!

    /// Shared URLSession configured with appropriate timeouts.
    private lazy var session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        config.timeoutIntervalForResource = 60
        return URLSession(configuration: config)
    }()

    /// Fetches a summarised daily digest for the given country.  The
    /// function constructs a prompt asking the model to research the
    /// most important stories for the country using its built‑in web
    /// search, summarise them and return structured JSON.  On success
    /// it returns the raw JSON string.  Errors are propagated.
    func fetchDailyDigest(for countryCode: String) async throws -> String {
        guard let apiKey = ProcessInfo.processInfo.environment["OPENAI_API_KEY"], !apiKey.isEmpty else {
            throw ConfigurationError.missingAPIKey
        }

        // Define the JSON schema we expect back from the model.  We
        // request a top level object with a daily_summary and an array
        // of stories.  Each story contains required fields.  For
        // details see OpenAI's structured output docs.
        let schema: [String: Any] = [
            "type": "object",
            "properties": [
                "country_code": ["type": "string"],
                "generated_at": ["type": "string"],
                "daily_summary": ["type": "string"],
                "stories": [
                    "type": "array",
                    "items": [
                        "type": "object",
                        "properties": [
                            "headline": ["type": "string"],
                            "source_name": ["type": "string"],
                            "source_url": ["type": "string"],
                            "category": ["type": "string"],
                            "summary": ["type": "string"],
                            "tldr": ["type": "string"],
                            "why_it_matters": ["type": "string"]
                        ],
                        "required": ["headline", "source_name", "source_url", "summary"]
                    ]
                ]
            ],
            "required": ["country_code", "generated_at", "daily_summary", "stories"]
        ]

        let requestDict: [String: Any] = [
            "model": "gpt-5-mini",
            "max_output_tokens": 1500,
            "tools": [
                ["type": "web_search_preview"]
            ],
            "text": [
                [
                    "format": [
                        "type": "json_schema",
                        "name": "latam_digest",
                        "schema": schema
                    ]
                ]
            ],
            "input": [
                [
                    "role": "system",
                    "content": [
                        [
                            "type": "input_text",
                            "text": "You prepare concise daily country news digests for a mobile app."
                        ]
                    ]
                ],
                [
                    "role": "user",
                    "content": [
                        [
                            "type": "input_text",
                            "text": "Gather the most important news stories for country code \(countryCode) from the past 24 hours. Deduplicate similar articles. For each story, provide the headline, source name, URL, category, summary, TLDR and why it matters. Also provide a daily summary for the country. Return valid JSON that matches the provided schema."
                        ]
                    ]
                ]
            ]
        ]
        let bodyData = try JSONSerialization.data(withJSONObject: requestDict, options: [])
        var request = URLRequest(url: baseURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.httpBody = bodyData
        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        return String(data: data, encoding: .utf8) ?? ""
    }
}

File: backend/src/openaiService.js
import OpenAI from "openai";
import { z } from "zod";
import { config } from "./config.js";
import { digestSchema } from "./schema.js";

const storySchema = z.object({
  headline: z.string(),
  source_name: z.string(),
  source_url: z.string().url(),
  category: z.string(),
  summary: z.string(),
  why_it_matters: z.string()
});

const digestResponseSchema = z.object({
  country_code: z.string(),
  country_name: z.string(),
  generated_at: z.string(),
  daily_summary: z.string(),
  stories: z.array(storySchema)
});

let client;

function getClient() {
  if (!config.openaiApiKey) {
    throw new Error("OPENAI_API_KEY is not configured on the backend.");
  }

  if (!client) {
    client = new OpenAI({ apiKey: config.openaiApiKey });
  }

  return client;
}

export async function generateDigest({ countryCode, countryName }) {
  const response = await getClient().responses.create({
    model: config.openaiModel,
    input: `Create a high-quality AI-curated daily news digest for ${countryName} (${countryCode}) focused on the last 24 hours. 

Requirements:
- Use web search to find credible recent reporting.
- Deduplicate overlapping stories.
- Include the most important political, economic, business, crime, sports, technology, culture, international, and society stories if relevant.
- Prefer primary reporting and major reputable outlets.
- Return 6 to 10 stories.
- Each summary should be concise and useful in a mobile app.
- "why_it_matters" should explain relevance in one short paragraph.
- Use the article URL from the reporting source, not a homepage.
- Output only data that matches the required JSON schema.`,
    instructions:
      "You are the editorial engine for a Latin America news app. Focus on accuracy, recency, and concise synthesis. Avoid sensationalism. If evidence is weak, omit the story instead of guessing.",
    tools: [{ type: "web_search" }],
    include: ["web_search_call.action.sources"],
    max_tool_calls: 8,
    max_output_tokens: 2200,
    text: {
      format: {
        type: "json_schema",
        name: "latam_digest",
        strict: true,
        schema: digestSchema
      }
    }
  });

  const parsed = digestResponseSchema.parse(JSON.parse(response.output_text));

  return {
    ...parsed,
    request_id: response._request_id,
    sources: extractSources(response)
  };
}

function extractSources(response) {
  const sources = [];

  for (const item of response.output || []) {
    if (item.type !== "web_search_call" || !item.action?.sources) {
      continue;
    }

    for (const source of item.action.sources) {
      sources.push({
        title: source.title || null,
        url: source.url || null
      });
    }
  }

  return dedupeSources(sources);
}

function dedupeSources(sources) {
  const seen = new Set();

  return sources.filter((source) => {
    const key = `${source.title || ""}|${source.url || ""}`;
    if (seen.has(key)) {
      return false;
    }

    seen.add(key);
    return true;
  });
}
